src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala

```diff
@@ -488,6 +488,31 @@ abstract class ExplicitOuter extends InfoTransform
             transform(treeCopy.Apply(tree, treeCopy.Select(eqsel, outerSelect, eq), args))
           }
 
+        // (t12312) C.this.a().X().isInstanceOf[C.this.a.X.type]() -->
+        // D.this.$outer().a().X().isInstanceOf[D.this.$outer.a.X.type]()
+        case TypeApply(fun, targs) =>
+          val rewriteTypeToExplicitOuter = new TypeMap { typeMap =>
+            def apply(tp: Type) = tp map {
+              case ThisType(sym) if sym != currentClass && !(sym.hasModuleFlag && sym.isStatic) =>
+                var cls = currentClass
+                var tpe = cls.thisType
+                do {
+                  tpe = singleType(tpe, outerAccessor(cls))
+                  cls = cls.outerClass
+                } while (cls != NoSymbol && sym != cls)
+                tpe.mapOver(typeMap)
+              case tp => tp.mapOver(typeMap)
+            }
+          }
+          val fun2   = transform(fun)
+          val targs2 = targs.mapConserve { targ0 =>
+            val targ    = transform(targ0)
+            val targTp  = targ.tpe
+            val targTp2 = rewriteTypeToExplicitOuter(targTp.dealias)
+            if (targTp eq targTp2) targ else transform(TypeTree(targTp2).setOriginal(targ))
```

Comment:
Why do you recurse here?

Reply:
Good question.  I tried to look around and I couldn't tell if I should recurse before or after my change.  Looks like here I opted for both.  And the reason for recursing is for completeness?  I.e. to avoid the top level transform not seeing this TypeTree (in some future case TypeTree).  Is there a rule of thumb on this?

Reply:
Here you create a `TypeTree`, does the transform do anything to that tree? I'd guess not. Also, the `targ` tree from with the type is derived is already transformed.